file/src/test/scala/docs/scaladsl/ArchiveSpec.scala

```diff
@@ -116,6 +117,97 @@ class ArchiveSpec
         archiveHelper.unzip(akkaZipped.futureValue).asScala shouldBe inputFiles
       }
     }
+
+    "tar flow is called" should {
+      "pass empty stream" in {
+        val sources = Source.empty
+        val tarFlow = Archive.tar()
+
+        val akkaTarred: Future[ByteString] =
+          sources
+            .via(tarFlow)
+            .runWith(Sink.fold(ByteString.empty)(_ ++ _))
+
+        akkaTarred.futureValue shouldBe ByteString.empty
+      }
+
+      "archive file" in {
+        val filePath1 = getPathFromResources("akka_full_color.svg")
+        val filePath2 = getPathFromResources("akka_icon_reverse.svg")
+        val fileStream1: Source[ByteString, Any] = FileIO.fromPath(filePath1)
+        val fileStream2: Source[ByteString, Any] = FileIO.fromPath(filePath2)
+        val fileContent1 = ByteString(Files.readAllBytes(filePath1))
+        val fileContent2 = ByteString(Files.readAllBytes(filePath2))
+        val fileSize1 = Files.size(filePath1)
+        val fileSize2 = Files.size(filePath2)
```

Comment:
I wonder if we can provide helpers to get the size from the stream, but that requires to process the source twice.

For example
```scala
            .map { path =>
              path.getFileName.toString -> FileIO.fromPath(path)
            }
            .mapAsync(1) {
              case (name, fileSource) =>
                val size = fileSource.runWith(Sink.fold(0)(_ + _.size))
                size.map { s =>
                  ArchiveMetadataWithSize(name, s) -> fileSource
                }(system.dispatcher)
            }
```

Reply:
Yeah it is unfortunate, that for TAR files you need to know the size of the files upfront. But I think  finding out the size is not in the scope of this.

But what I maybe should add is a check, that one actually provides a stream of matching size. So Running

```
Source((ArchiveMetadataWithSize("file.txt", 1), Source.single(ByteString("too long"))).runWith(...)
```

should fail.

Reply:
@ennru What you think about this for verifying the actual stream size matches the header:

```scala
  private def ensureFileSize(header: TarArchiveHeader): Flow[ByteString, ByteString, NotUsed] =
    Flow[ByteString]
    .map(Option.apply)
    .concat(Source.single(None))
    .statefulMapConcat(() => {
      var size = 0L
      elem => elem match {
        case Some(chunk) =>
          size = size + chunk.size
          chunk :: Nil
        case None =>
          if (size == header.size) Nil
          else throw new IllegalStateException(s"Byte stream for ${header.filePath} had $size bytes but expected were ${header.size} bytes")
      }
    })
```

Reply:
Wrote the size checking as a custom stage (so I don't have to wrap every single chunk in a `Some(_)`).

Reply:
That's a great addition.
In many situations, the size might be known. Let's leave the auto-detection for now.